lib/spack/spack/compiler.py

```diff
@@ -655,6 +667,65 @@ def search_regexps(cls, language):
             for prefix, name, suffix in itertools.product(prefixes, compiler_names, suffixes)
         ]
 
+    @classmethod
+    def _cmp_test_env(cls, cmp_pth):
+        return os.environ
+
+    @classmethod
+    def is_supported_on_platform(cls, plat):
+        return type(plat) in cls.supported_platforms
+
+    @overload
+    def try_compiler(self, lang, file=None, flags=[], definitions=[]):
+        cmp_pth = getattr(self, lang)
+        if cmp_pth:
+            return self.try_compiler(cmp_pth, file, lang, flags, definitions)
+        raise TryCompilerError(self.spec.name, file, lang, f"No compiler for lang {lang}")
+
+    @overload
+    @classmethod
+    def try_compiler(cls, compiler_path, file=None, lang=None, flags=[], definitions=[]):
+        """Try to use this compiler to compile a test file of specified language.
+        File is provided by caller or specified by language denotation.
+        Args:
+            compiler_path (str): path to compiler to be tested
+            file (str): path to file to be test compiled
+        Returns output of compilation if successful, Raises TryCompilerError otherwise
+        """
+
+        # if file is not set by caller, use default file for specified language
+        # if there is no specified language, we don't know what to compile, fail
+        determinable_file = file or lang
+        if not determinable_file:
+            err = "No file or language provided."
+            raise TryCompilerError(cls.__name__, file, lang, err)
+
+        test_file = file if file else default_try_file_for_lang(lang)
```

Comment:
We'll want to replicate the logic from `build_environment.setup_package` that ensures that all env vars that need to be defined for the compiler are set - that includes:

- `set_compiler_environment_variables`
-  `for mod in pkg.compiler.modules: load_module(mod)`

(and this should happen in a subprocess, or a context manager that undoes env var modifications).

Reply:
Good point.
I think `Compiler.compiler_environment` will handle the environment context as well as a good bit of the setup (it calls `load_module` under the hood), I'll just need to make sure we cover the rest of the setup.